### PR TITLE
Inefficient Usage of Java Collections

### PR DESCRIPTION
**Pull request by FastAtlas**
Hi,

We find that there are several ArrayList objects which are not manipulated by random access. Due to the memory reallocation triggered in the successive insertions, the time complexity of add method of ArrayList is amortized O(1). We notice that these objects are only used for traversal, the retrieval, and removal of the first or the last element.

This functionality can be implemented by LinkedList. Moreover, the insertion of LinkedList is strictly O(1) time complexity because no memory reallocation occurs.

Meanwhile, we also found several TreeMap and TreeSet objects which are not necessary to maintain the order of keys. To achieve the same functionality, HashMap and HashSet are enough. The replacement can also reduce the time cost in the modification of the set and the map objects. A similar case is the replacement of LinkedHashMap when it is not iterated actually. HashMap is also sufficient.

Also, there is no need to change the set to the list in the function `extendsKeywords` in ExtendedKeywords.java because the index-value correlation does not matter in further use. Only the iterators and the insertions are invoked to the list. We can simply maintain the content by a HashSet, and the contains method is more efficient than the one of List.

We discovered the above inefficient usage of containers by our tool Ditto. The patch is submitted. Could you please check and accept it? We have tested the patch on our PC. The patched program works well.

Bests

Ditto

### CODE REVIEW

**Comment by Ladicek** [comment]
I won't comment on the other changes, but I'd like to highlight one thing about the `ArrayList` -> `LinkedList` move.

The advantages of linked lists over arrays in scenarios where the list is only iterated or appended are only real under the RAM computation model. Real computers are not RAM computers, they employ memory hierarchies, and so the same logic does _not_ apply. Using a `LinkedList` is almost always _worse_ than using an `ArrayList`.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/reflection/GrpcServerIndex.java [hunk 1]
```diff
@@ -33,9 +33,9 @@ public GrpcServerIndex(
         Set<String> files = new HashSet<>();
         Set<String> names = new HashSet<>();
         Set<FileDescriptor> services = new HashSet<>();
-        Map<String, FileDescriptor> descriptorsByName = new LinkedHashMap<>();
-        Map<String, FileDescriptor> descriptorsBySymbol = new LinkedHashMap<>();
-        Map<String, Map<Integer, FileDescriptor>> descriptorsByExtensionAndNumber = new LinkedHashMap<>();
+        Map<String, FileDescriptor> descriptorsByName = new HashMap<>();
+        Map<String, FileDescriptor> descriptorsBySymbol = new HashMap<>();
+        Map<String, Map<Integer, FileDescriptor>> descriptorsByExtensionAndNumber = new HashMap<>();
```
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/reflection/GrpcServerIndex.java:38]: Insertion order was important. I need to see if that still the case. @michalszynkiewicz

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanProcessor.java [hunk 2]
```diff
@@ -170,7 +171,7 @@ public void processValidationErrors(BeanDeploymentValidator.ValidationContext va
                 injectionPointAnnotationsPredicate, allowMocking);
         AnnotationLiteralGenerator annotationLiteralsGenerator = new AnnotationLiteralGenerator(generateSources);
 
-        List<Resource> resources = new ArrayList<>();
+        List<Resource> resources = new LinkedList<>();
```
Comment (cescoffier) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanProcessor.java:174]: Doubting about the benefits of this change.

**Comment by DittoTool** [comment]
> I won't comment on the other changes, but I'd like to highlight one thing about the `ArrayList` -> `LinkedList` move.
> 
> The advantages of linked lists over arrays in scenarios where the list is only iterated or appended are only real under the RAM computation model. Real computers are not RAM computers, they employ memory hierarchies, and so the same logic does _not_ apply. Using a `LinkedList` is almost always _worse_ than using an `ArrayList`.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/reflection/GrpcServerIndex.java (hunk 1 above)
Comment (DittoTool) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/reflection/GrpcServerIndex.java:38]: I think these LinkedHashMap objects are not traversed in the program. As I have seen, only containsKey() and get() methods are invoked.

**Comment by gsmet** [comment]
My problem with this sort of PR is that it will make backporting things harder for very little benefits.

If we can find critical things that need to be changed and bring a significant improvement, let's do it. But otherwise I would just let it at that.

**Comment by Sanne** [comment]
> OK. I got your point. Theoretically, such replacements can improve efficiency. 

In theoretical computer science classes this is true.

> [..] in the real computer, the complexity specification might not hold. 

To be more precise, we do know for sure that LinkedLists perform worse when working in Java. This is not a "might" fact but something that has been investigated carefully.

I do appreciate the attempts to improve efficiency though!  :+1:  But when making performance related contributions, please measure  things carefully.

**Pull request closed without merging**